components/brave_wallet/browser/eth_allowance_manager.cc

```diff
@@ -0,0 +1,409 @@
+    OnDiscoverEthAllowancesCompleted({});
+    return;
+  }
+
+  const auto token_list_map =
+      BlockchainRegistry::GetInstance()->GetEthTokenListMap(
+          GetChainIdsForAlowanceDiscovering());
+
+  if (!allowance_discovery_tasks_.empty()) {
+    allowance_discovery_tasks_.clear();
+    get_block_tasks_ = 0;
+  }
+  for (const auto& [chain_id, token_list] : token_list_map) {
+    base::Value::List contract_addresses;
+    for (const auto& token : token_list) {
+      contract_addresses.Append(token->contract_address);
+    }
+
+    if (contract_addresses.empty()) {
+      continue;
+    }
+    auto internal_callback = base::BindOnce(
+        &EthAllowanceManager::OnGetCurrentBlock, weak_ptr_factory_.GetWeakPtr(),
+        chain_id, std::move(contract_addresses), account_addresses);
+    get_block_tasks_++;
+    json_rpc_service_->GetBlockNumber(chain_id, std::move(internal_callback));
+  }
+}
+
+void EthAllowanceManager::OnGetCurrentBlock(
+    const std::string& chain_id,
+    base::Value::List contract_addresses,
+    const std::vector<std::string>& account_addresses,
+    uint256_t block_num,
+    mojom::ProviderError error,
+    const std::string& error_message) {
+  const std::string latest_block_num_hex(error == mojom::ProviderError::kSuccess
+                                             ? Uint256ValueToHex(block_num)
+                                             : kLatestBlock);
+  const auto& allowance_cache_dict =
+      prefs_->GetDict(kBraveWalletEthAllowancesCache);
+
+  auto get_blocknumber_filter_from_cache =
+      [&](const std::string& chain_id, const std::string& hex_account_address) {
+        std::string result(kEarliestBlock);
+        const auto* chain_cached_data = allowance_cache_dict.FindDict(chain_id);
+        if (!chain_cached_data) {
+          return result;
+        }
+
+        const auto* last_block_number_section =
+            chain_cached_data->FindDict(kLastBlockNumber);
+        if (!last_block_number_section) {
+          return result;
+        }
+
+        const auto* last_block_number_ptr =
+            last_block_number_section->FindString(hex_account_address);
+        if (!last_block_number_ptr) {
+          return result;
+        }
+
+        uint256_t block_number{0};
+        if (!last_block_number_ptr->empty() &&
+            HexValueToUint256(*last_block_number_ptr, &block_number)) {
+          result = Uint256ValueToHex(++block_number);
+        }
+        return result;
+      };
+  const auto approval_topic_hash = KeccakHash(kApprovalTopicHash);
+  for (const auto& account_address : account_addresses) {
+    std::string account_address_hex;
+    if (!PadHexEncodedParameter(account_address, &account_address_hex)) {
+      continue;
+    }
+    const int task_id(allowance_discovery_tasks_.size());
+    auto internal_callback = base::BindOnce(
+        &EthAllowanceManager::OnGetAllowances, weak_ptr_factory_.GetWeakPtr(),
+        task_id, chain_id, account_address_hex);
```

Comment:
chain_id and account_address_hex already part of task item, no need to bind them.

Reply:
fixed